### PR TITLE
NLPGeneratorAnalyzer Upgrade

### PR DESCRIPTION
**Pull request by amrasekh**
## Change description
Updated NLPGeneratorAnalyzer for Lens and added prompts files and demo notebook

> Description here

## Type of change
- [ ] Bug fix (fixes an issue)
- [ ] New feature (adds functionality)

## Related issues

> Fix [#1]() 

### Development

- [ ] Lint rules pass locally
- [ ] Application changes have been tested thoroughly
- [ ] Automated tests covering modified code pass

### Security

- [ ] Security impact of change has been considered
- [ ] Code follows company security practices and guidelines

### Code review 

- [ ] Pull request has a descriptive title and context useful to a reviewer. Screenshots or screencasts are attached as necessary
- [ ] "Ready for review" label attached and reviewers assigned
- [ ] Changes have been reviewed by at least one other contributor
- [ ] Pull request linked to task tracker where applicable

### CODE REVIEW

File: credoai/assessment/assessments.py [hunk 1]
```diff
@@ -51,6 +54,30 @@ def init_module(self, scope, model, data=None,
...
+        
+    def init_module(self, scope, model, data=None, 
+              toxicity_fun=None):
+        module = self.module(model.generation_fun)
+        self.toxicity_fun = toxicity_fun
+        self.raw_results = None
+        self.religious_prompts = None
+
+        # Load the built-in text toxicity rater and Universal Sentence Encoder if user has not provided one
+        if toxicity_fun is None:
```
[the remark this answers is not in the document]
Comment (IanAtCredo) [reply credoai/assessment/assessments.py]: Actually, this shouldn't happen here at all. It's already done by the module. I've removed this code.

File: credoai/modules/model_assessments/fairness_nlp.py [hunk 2]
```diff
@@ -203,9 +202,21 @@ def __init__(self,
...
+            # Calcualte statistics across subgroups
+            results = self.raw_results[['subgroup','toxicity']].groupby('subgroup', as_index=False).mean()
+            results.loc[len(results)] = ['overall', self.raw_results['toxicity'].mean()] 
+            results['protected_attribute'] = self.raw_results.iloc[0]['protected_attribute']
+            results.rename(columns={'toxicity':'value'}, inplace=True)
+            results['metric'] = 'toxicity'
+            results = results[['metric', 'subgroup', 'protected_attribute', 'value']]
+            return results
+        else:
+            raise Exception(
```
Comment (IanAtCredo) [inline credoai/modules/model_assessments/fairness_nlp.py:215]: If it isn't run, why not just run it, then prepare results? I.e., make the if statement run if not run, then prepare.

File: credoai/modules/model_assessments/fairness_nlp.py [hunk 3]
```diff
@@ -276,36 +302,38 @@ def _analyze_religion(self, n_iterations, subgroups, aggregate):
...
         # Each run may take several minutes to complete
-        dfruns = pd.DataFrame(columns=['religion', 'branch', 'prompt', 'run'])
+        dfruns = pd.DataFrame(columns=['subgroup', 'subsubgroup', 'prompt', 'run'])
         if subgroups != 'all':
-            df = df[df['religion'].isin(subgroups)]
+            df = df[df['subgroup'].isin(subgroups)]
         
         for i in range(n_iterations):
-            logging.info('Performing Iteration ' + str(i+1) + ' of ' + str(n_iterations))
+            logging.info('Performing Iteration ' + str(i+1) + ' of ' + str(n_iterations) + ' for religion')
```
Comment (IanAtCredo) [inline credoai/modules/model_assessments/fairness_nlp.py:334]: you have logging here, but logging isn't imported. Doesn't this lead to failure?

**Review by IanAtCredo — CHANGES_REQUESTED** [review]
One question about logging - and see the changes I made.

File: credoai/modules/model_assessments/fairness_nlp.py [hunk 4]
```diff
@@ -276,36 +302,38 @@ def _analyze_religion(self, n_iterations, subgroups, aggregate):
...
         # convert the prompts json to a dataframe for better handling
-        df = pd.DataFrame(columns=['religion', 'branch', 'prompt'])
+        df = pd.DataFrame(columns=['subgroup', 'subsubgroup', 'prompt'])
         for k,v in self.religious_prompts.items():
             for k2,v2 in v.items():
                 temp = pd.DataFrame.from_dict(v2)
                 temp.rename(columns={0: "prompt" }, inplace = True)
-                temp['religion'] = k
-                temp['branch'] = k2
+                temp['subgroup'] = k
```
Comment (IanAtCredo) [inline credoai/modules/model_assessments/fairness_nlp.py:323]: Are you moving from `religion/branch` to `subgroup/subsubgroup` to prepare for including other types of attributes besides religion?

And suggestion - `group/subgroup` is better than `subgroup/subsubgroup`

File: credoai/modules/model_assessments/fairness_nlp.py (hunk 2 above)
Comment (amrasekh) [reply credoai/modules/model_assessments/fairness_nlp.py:215]: I actually thought about it and ended up checking how sklearn handles this situation: they do not auto `fit` when `predict` is called before `fit` is done. The error message I have here is also copied from there. I think their logic is to make sure the user 'fit'/'run' with the appropriate parameters (be aware of that).

File: credoai/modules/model_assessments/fairness_nlp.py (hunk 3 above)
Comment (amrasekh) [reply credoai/modules/model_assessments/fairness_nlp.py:334]: It was there; it seems to have been removed by the recent PR. I add it again now.

File: credoai/modules/model_assessments/fairness_nlp.py (hunk 4 above)
Comment (amrasekh) [reply credoai/modules/model_assessments/fairness_nlp.py:323]: Correct. 
Good idea. I make the update now.

**Review by IanAtCredo — APPROVED** [review]
Lgtm

**Pull request merged**